### DISCUSSION TITLE
Extending group permissions to subgroups

### DISCUSSION THREAD

**Discussion by antikalk** [keycloak/keycloak#16915] [Ideas]
Hi everyone,

I am currently trying out some things with the fine grained authorization for groups and found out that it seems it is currently not possible to inherit permissions to child groups. Example:

1. Create a user with role `query-groups`
2. Create group with name `parent`
3. Create subgroup of group `parent` with the name child
4. Now enable permissions for group `parent`
5. Add a user policy for the created user
6. Add this policy to the `view` permission of the `parent` group
7. Login in the admin-ui as the user

You will then be able to view the parent group but will receive a 403 error when trying to access the child group. Same thing applies for the other permissions (e.g. view-members -> with the `query-users` role added to the user I will only see users of the parent role, but not of the child role in the users overview). Only workaround would be to explicitly add permissions to every group. 

In my opinion group permissions should be extended to the children. 

You are also able to see all groups in the overview, not only the once you have the view permission for (maybe a filter should be added here?)

**Comment by antikalk** [comment]
I created a bug for the `view-members` of child groups permission, as there is already code that should actually support this - which does not work on my side: https://github.com/keycloak/keycloak/issues/16966

Still it would be nice if this inheritance would somehow apply for all group related permissions :)

**Comment by ssilvert** [comment]
I think this would become extremely complicated if we allowed inheritance.  Also, current users may be relying on the current behavior.  So inheritance would have to be optional.

**Reply by thw0rted** [reply]
We are running into a similar problem now.  The issue is we would like to allow a non-admin client to "own" a group and be able to freely create and manage subgroups under it.  With the current model, fine-grained permissions can be configured to allow this client to create subgroups, but since the child is created without fine-grained permissions enabled, and the client in question does not have realm-level privileges to configure fine-grained permissions globally, the newly created subgroup is effectively "lost" for that client -- they can see it as a subgroup when querying the parent, but they cannot query the subgroup, configure permissions on it, add/remove members, etc.

I understand that allowing inheritance would create additional complexity, but what about automatically cloning the parent's fine-grained permissions when creating subgroups?  Wouldn't that be a sensible default?  This would allow a client with permissions on `parent` create a `child` where they have the same permissions to start, without requiring them to have realm-level permission-management privileges.

**Reply by antikalk** [reply]
@thw0rted imo this would be a possibility, but would create a lot of (maybe unnecessary) copies of the fine-grained permission settings. Also syncing changes on the parent would then again be complicated... and you would need a toggle during the creation of the subgroup whether you wanna create those settings or not.

Coming back to my intial idea and the feedback from @ssilvert  it could be implemented like this: 
- Add "inheritPermissionsFromParent" boolean flag on Group Entity. If set to true parent fine-grained permissions of this group are evaluated if the group itself does not have any for the scope requested
- During the evaluation of the permission on a subgroup, if the group does not have his own permissions configured for the scope requested, check for the parents if the flag is set (very pseudo code - just to demonstrate the idea):
```
const scope = 'manage_members'; // example
if (group.hasFineGrainedPermissionSettings(scope) {
  // evaluate group fine grained permission ...
} else if (group.inheritPermissionsFromParent()) {
  // check parents fine graind permissions
  let parent = group.getParent();
  while (parent != null) {
     if (parent.hasFineGrainedPermissionSettings(scope)) {
       // evaluate parent fine grained permission and return ...
     } else if (parent.inheritPermissionsFromParent()) {
       // check the grandparent
       parent = parent.getParent();
     } else {
       parent = null;
     }
  }
}
// return access denied
```
But as there is a discussion pending for [Advanced Groups](https://github.com/keycloak/keycloak/discussions/19055) I am not sure if this topic is still that relevant...

**Reply by oculos** [reply]
Yes! This solution - an optional inheritance of the parents' permissions - would be great! We are dealing with a possibility to let some users administer a group, create subgroups and populate them, but the problem is that, everytime they create a child group, an admin needs to copy permissions there.

**Reply by LevN0** [reply]
Seconding. Optional inheritance of the parents' permissions would indeed be great.

**Reply by MartaGolabek** [reply]
Optional inheritance would be absolutely essential for me as well.

**Comment by rkdutta** [comment]
I am evaluating the fine-grain-admin-authz feature. My Keycloak version is 26.1.4.
I still find it challenging to extend parent group permission to child group. For each child group, I need to explicitly create a permission. Is this already possible to inherit permission from parent to child group? Could you please guide?

Note: I have also mentioned the same comment [here](https://github.com/keycloak/keycloak/issues/17588#issuecomment-2745166568)

**Reply by LevN0** [reply]
This is still not supported as of keycloak 26.4. I opened a feature enhancement on it in #43865.

**Reply by dsc-mgl** [reply]
@LevN0 Do you know whether this is on the roadmap anytime soon?
